docs/design/downward_api_resources_limits_requests.md

```diff
@@ -0,0 +1,530 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should refer to
+the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Downward API for resource limits and requests
+
+## Background
+
+Currently the downward API (via environment variables and volume plugin) only
+supports exposing a Pod's name, namespace, annotations, labels and its IP
+([see details](http://kubernetes.io/docs/user-guide/downward-api/)). This
+document explains the need and design to extend them to expose resources
+(e.g. cpu, memory) limits and requests.
+
+## Motivation
+
+Software applications require configuration to work optimally with the resources they're allowed to use.
+Exposing the requested and limited amounts of available resources inside containers will allow
+these applications to be configured more easily. Although docker already
+exposes some of this information inside containers, the downward API helps
+exposing this information in a runtime-agnostic manner in Kubernetes.
+
+## Design
+
+This is mostly driven by the discussion in [this issue](https://github.com/kubernetes/kubernetes/issues/9473).
+There are three approaches discussed in this document to obtain resources limits
+and requests to be exposed as environment variables and volumes inside
+containers:
+
+1. The first approach requires users to specify full json path selectors
+in which selectors are relative to the pod spec. The benefit of this
+approach is to specify pod level resources, and since containers are
+also part of a pod spec, it can be used to specify container level
+resources too.
+
+2. The second approach requires specifying partial json path selectors
+which are relative to the container spec. This approach helps
+in retrieving a container specific resource limits and requests, and at
+the same time, it is simpler to specify than full json path selectors.
+
+3. In this approach, users specify particular strings to retrieve
```

Comment:
Yes. As per my understanding, these magic strings are treated just as strings not json path (though they "look" same as json paths) in current downward API implementation that means strings (or parts of the strings) are not processed (and at the same time not validated) against the actual json tags in the versioned kube objects. The similar approach is used in the 3rd approach in this proposal, and thats why I call them no json path selectors, again even though they "look" same as json path.  

To process strings as json paths, one would have to use go reflections (I am not sure if there is any other way) as is done in json path library and is used in kubectl. One example is that if lets say If change json tags `metadata` and `name` to something like `metapart` and `realname`, the magic string "metadata.name" would keep working in the current downward API, but will fail in kubectl as `metadata.name` is incorrect now and the correct one is `metapart.realname` . 

Please understand I am `Not` saying the current downward API is wrong (the current approach is perfectly fine), I am just saying it does not process json paths as json paths but as magic strings as is the 3rd approach, but is very different from the 1st and 2nd approaches.